neps/nep-0371.md

```diff
@@ -0,0 +1,149 @@
+
+## Rationale and alternatives
+
+Without creating a promise, we can only query for information stored on the same shard. This is why in this proposal, the access key information at runtime will only come from keys stored on the current contract. We had also investigated the possibility of introducing pagination for all keys stored on the contract but this is infeasible at the moment due to the significant redesign requirement of the VM logic.
+
+## Runtime specification
+
+The contract runtime provides one host function called `access_key_info` which is callable in regular calls and views.
+
+```rust
+fn access_key_info (
+    public_key_len: u64,                // in:  number of bytes of the public key to query
+    public_key_ptr: u64,                // in:  pointer to the public key to query
+    nonce_ptr: u64                      // out: pointer to where a u64 can be written by the host
+    allowance_ptr: u64,                 // out: pointer to where a u128 can be written by the host
+    account_id_register_id: u64,        // out: register to write the access key holder account id
+    method_names_register_id: u64,      // out: register to write the method names as a comma separated list
+)
+-> u64;
+```
+
+### Return values
+
+- Returns 0 if the given public key is not registered as access key to the current account. In this case, the memory locations behind `nonce_ptr` and `allowance_ptr` are not changed by the host. The register content of `account_id_register_id` and `method_names_register_id` are not changed, either.
+- Returns 1 if the given public key is a full access key to the current account. The memory location at `nonce_ptr` contains the current nonce value of the access key. The memory location at `allowance_ptr` is not changed by the host. The register content of `account_id_register_id` and `method_names_register_id` are not changed, either.
+- Returns 2 if the given public key is a function call access key to the current account. The memory location at `nonce_ptr` contains the current nonce value of the access key. The memory location at `allowance_ptr` contains a `u128` value that is the remaining allowance in yoctoNEAR. The register `account_id_register_id` contains the account name of the access key holder. The register `method_names_register_id` contains a comma separated list of all method names that the access key is valid for.
+
+### Panics
+
+- `InvalidPublicKey`: Calling `access_key_info` with an invalid public key.
```

Comment:
Just for my understanding, how would this error be raised?  I would've expected that the implementation will iterate over the list of keys registered against the contract and compare them against the one that the contract passed in.  This error seems to suggest that we need to do more processing on the key that is passed in.

Reply:
This is essentially a deserialization error for a `PublicKey` struct. Simplest example would be an invalid length.

Regarding implementation, we will not iterate the list of keys. Each access key is stored in the trie at a specific trie key. The trie key is defined as `col::ACCESS_KEY ++ account_id ++ col::ACCESS_KEY ++ public_key`. So we will directly look up this node in the trie.